### PR TITLE
Fix: Performance: Batch multiple activation function evaluations in single GPU call (#201)

### LINKED ISSUE #201: Performance: Batch multiple activation function evaluations in single GPU call
## Summary
Reduce GPU round-trips by batching multiple activation function evaluations (TANH, SIGMOID, ReLU, etc.) into a single GPU call instead of separate calls per activation type.

## Current Behaviour
- `ACTIVATION_SPECS` are evaluated sequentially (lines 8239-8260 in implementation.rs)
- Each activation type requires a separate GPU buffer submission
- For neuron candidates, this means N GPU round-trips where N = number of activation types tested

## Proposed Solution
Create a combined shader that evaluates multiple activation functions in parallel:

### Option A: Multi-output shader
```wgsl
// Single shader that outputs results for all activation types
@compute @workgroup_size(256)
fn evaluate_all_activations(...) {
    // For each sample:
    // - Compute TANH improvement
    // - Compute SIGMOID improvement  
    // - Compute ReLU improvement
    // - etc.
    // Store all results in output buffer
}
```

### Option B: Batched single-output shader
- Submit all activation evaluations in one command buffer
- Map all result buffers together
- Process results in single CPU pass

## Expected Impact
- **10-20% fewer GPU round-trips** during neuron analysis
- Reduced CPU-GPU synchronization overhead
- Better GPU utilization (larger, fewer batches)

## Implementation Notes
- Current shaders are in `src/shaders/` (activation.wgsl, bias.wgsl, relu.wgsl)
- Must maintain current workgroup size (256) for Apple Silicon SIMD efficiency
- Consider activation-specific paths vs unified path (performance vs code complexity)

## Risks
- Shader divergence: different activation functions have different math, may cause GPU pipeline stalls
- Code complexity: combined shader is harder to maintain
- Testing: need to verify correctness for all 19 activation types

## Testing Strategy (TDD)
1. Create test comparing batched vs sequential results (must be identical)
2. Verify all 19 activation functions work correctly in batched mode
3. Test edge cases (saturation, negative inputs, etc.)

## Benchmark Requirements
- Measure GPU round-trips before/after
- Measure total neuron analysis time before/after
- Profile GPU utilization on M4 and older Apple Silicon

## Labels
`performance`, `enhancement`, `gpu`

---
*Created in response to #184*

### PR DESCRIPTION
**Pull request by stservice**
## Summary

Issue #201: Implemented batched activation function evaluation to reduce GPU round-trips during neuron analysis.

### Changes

1. **New `GpuEvaluator` trait method**: Added `evaluate_activations_batched()` to evaluate multiple activation function configurations in a single GPU command buffer submission.

2. **GPU implementation** (`src/analysis/gpu/analyzer.rs`):
   - `evaluate_activations_batched_gpu()` uploads sample data once, creates all compute passes in a single command buffer, and maps all result buffers together
   - Eliminates per-config GPU round-trip overhead

3. **GpuWorkQueue support** (`src/analysis/gpu/queue.rs`):
   - Added `ActivationBatchEval` work request variant
   - Thread-safe batched evaluation via the work queue

4. **Neuron analysis integration** (`src/analysis/neuron.rs`):
   - Replaced sequential `for spec in ACTIVATION_SPECS.iter()` loop with single call to `evaluate_all_activation_specs_batched()`
   - Maintains identical results while dramatically reducing GPU calls

5. **Helper function** (`src/analysis/synapse.rs`):
   - `evaluate_all_activation_specs_batched()` collects all (activation_type, orientation, scale) combinations for all 15 activation specs
   - Falls back to sequential evaluation if batched fails

### Performance Results

Benchmark comparing batched vs sequential evaluation on Apple M4 Pro:

```
Issue #201 Performance Comparison:
  Sample count: 500
  Activation configs: 256 (15 specs × orientations × scales)

Results (average of 5 iterations):
  Sequential: 7,054,620µs (256 GPU calls)
  Batched:    69,103µs (1 GPU call)
  Speedup:    102.09x
  Time reduction: 99.0%
```

**Key findings:**
- **102x speedup** (far exceeding the expected 10-20% improvement)
- Reduced from 256 GPU round-trips to just 1
- Same sample data uploaded only once, shared across all activation configs

### Why the dramatic improvement?

The original implementation made separate GPU calls for each (activation_type, orientation, scale) combination:
- 15 activation specs × 2 orientations × 8-9 scales ≈ 256 GPU calls per (source, target) pair
- Each call involved: buffer creation → command submission → buffer mapping → result readback

The batched implementation:
- Creates all buffers upfront
- Submits all compute passes in a single command buffer
- Maps all staging buffers concurrently
- Processes all results in one pass

### Unit Tests (`tests/issue_201_batched_activation.rs`)
- `test_batched_vs_sequential_results_identical` - Verifies batched produces identical results to sequential
- `test_all_activation_specs_in_batched_mode` - Verifies all 15 activation specs work correctly
- `test_batched_evaluation_empty_samples` - Edge case: empty sample input
- `test_batched_evaluation_single_config` - Edge case: single activation config
- `test_batched_evaluation_empty_configs` - Edge case: empty config input
- `test_batched_evaluation_edge_case_samples` - Edge case: NaN/Infinity values

### Performance Test (`tests/issue_201_benchmark.rs`)
- `test_batched_vs_sequential_performance` - Measures and reports speedup

### Benchmark (`benches/batched_activation.rs`)
- Criterion benchmark for accurate performance measurement

### All Existing Tests Pass
- `./quality.sh` passes with all 372 unit tests and 60+ integration tests

---

## Automated Fix Details
This PR addresses issue #201: **Performance: Batch multiple activation function evaluations in single GPU call**

## Quality Checks
- Followed TDD methodology
- All new functionality has corresponding tests
- Ran `./quality.sh` - all checks pass

## Notes
- No existing tests were removed or commented out
- If any test modifications were required due to business logic changes, they are documented in the commits

Closes #201

### CODE REVIEW

**Comment by stservice** [comment]
Thanks for the dependency review! ✅

No action required - the automated scan confirms that this PR:
- Has no vulnerabilities
- Has no license issues
- Has no OpenSSF Scorecard issues

All quality checks continue to pass (`./quality.sh` runs successfully with all 372 unit tests and 60+ integration tests passing).

File: tests/issue_201_benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+            }
+        }
+    }
+    configs
+}
+
+/// Performance comparison test.
+/// This test demonstrates the performance improvement from batching.
+#[test]
+fn test_batched_vs_sequential_performance() {
```
Comment (nleck) [inline tests/issue_201_benchmark.rs]: Shouldn't benchmarks be seperate to unit test ? Unit tests should check functionality not performance. 

In DenoJS they are very different. We run benchmarks singularly so that we get reliable perfomrnace numbers. We run unit tests in parrellel as we don't care so much about the performance but just the correct answer

**Review by nleck — COMMENTED** [review]
Benchmarks are NOT unit tests

**Comment by stservice** [comment]
You're right - I've removed the performance test from the unit tests.

**Changes made:**
- Deleted `tests/issue_201_benchmark.rs` which was measuring performance (speedup comparisons)

**What remains:**
- `benches/batched_activation.rs` - Criterion benchmark for accurate performance measurement (run with `cargo bench`)
- `tests/issue_201_batched_activation.rs` - Unit tests that verify **functionality** (correctness, not speed)

The unit tests now only check that:
- Batched and sequential produce identical results
- All activation specs work correctly in batched mode
- Edge cases are handled (empty samples, empty configs, NaN/Infinity values)

Performance validation stays in benchmarks where tests run singularly for reliable numbers.

**Review by nleck — APPROVED** [review]
